Sources/SwiftWarningControl/WarningControlDeclSyntax.swift

```diff
@@ -0,0 +1,104 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+/// Protocol representing a declaration scope which can potentially carry warning
+/// diagnostic group behavior controls (`@warn` attributes).
+protocol WarningControlDeclSyntax {
+  var attributes: AttributeListSyntax { get }
+  var position: AbsolutePosition { get }
+  var endPosition: AbsolutePosition { get }
+}
+
+// Compute a dictionary of all `@warn` diagnostic group behaviors
+// specified on this warning control declaration scope.
+extension WarningControlDeclSyntax {
+  func allWarningGroupControls() -> [String: WarningGroupBehavior] {
+    return attributes.reduce(into: [String: WarningGroupBehavior]()) { result, attr in
+      if case .attribute(let attributeSyntax) = attr,
```

Comment:
You could change all of the `if`s in this function to `guard` statements to avoid some nesting level.